File: #145_Binary_Tree_Postorder_Traversal/Solution.java
/* A solution using explicit stack (iterative) */

import java.util.*;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

public class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
    	List<Integer> result = new LinkedList();
    	if (root == null) return result;
    	Deque<TreeNode> stack = new LinkedList();
    	while (root != null || !stack.isEmpty()) {
    		if (root != null) {
	    		result.add(root.val);
    			stack.push(root);
    			root = root.right;
    		}
    		else {
    			root = stack.pop();
    			root = root.left;
    		}
    	}
    	Collections.reverse(result);
    	return result;
    }
}

File: #145_Binary_Tree_Postorder_Traversal/Solution.py
'''
Recursive Solution
'''

class TreeNode:
	def __init__(self, x):
		self.val = x
		self.left = None
		self.right = None

class Solution:

	def __init__(self):
		self.re = None

	# @param {TreeNode} root
	# @return {integer[]}
	def postorderTraversal(self, root):
		self.re = []
		self.postorder(root)
		return self.re

	def postorder(self, root):
		if not root:
			return
		self.postorder(root.left)
		self.postorder(root.right)
		self.re.append(root.val)